### PR TITLE
Add `needletail`

### PR DESCRIPTION
**Pull request by apcamargo**
This PR adds the [`needletail`](https://github.com/onecodex/needletail) Python library.

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

### General instructions

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

### Instructions for avoiding API, ABI, and CLI breakage issues
Conda is able to record and lock (a.k.a. pin) dependency versions used at build time of other recipes.
This way, one can avoid that expectations of a downstream recipe with regards to API, ABI, or CLI are violated by later changes in the recipe.
If not already present in the meta.yaml, make sure to specify `run_exports` (see [here](https://bioconda.github.io/contributor/linting.html#missing-run-exports) for the rationale and comprehensive explanation).
Add a `run_exports` section like this:

```yaml
build:
  run_exports:
    - ...

```

with `...` being one of:

| Case                             | run_exports statement                                               |
| -------------------------------- | ------------------------------------------------------------------- |
| semantic versioning              | `{{ pin_subpackage("myrecipe", max_pin="x") }}`     |
| semantic versioning (0.x.x)      | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}`   |
| known breakage in minor versions | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| known breakage in patch versions | `{{ pin_subpackage("myrecipe", max_pin="x.x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| calendar versioning              | `{{ pin_subpackage("myrecipe", max_pin=None) }}`    |

while replacing `"myrecipe"` with either `name` if a `name|lower` variable is defined in your recipe or with the lowercase name of the package in quotes.

### Bot commands for PR management

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

Note that the <code>@BiocondaBot please merge</code> command is now depreciated. Please just squash and merge instead.

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

### CODE REVIEW

**Comment by apcamargo** [comment]
@mbhall88 sorry for tagging you here, but I know you're working on better Rust documentation for Bioconda and we still don't have good recommendations for Maturin. Do you know what could be done here? Apparently, this approach of building a wheel and then installing it [doesn't support cross-compilation](https://github.com/conda-forge/polars-feedstock/issues/96). However, `pip` is not working for this package either.

**Comment by mbhall88** [comment]
@apcamargo I don't have any Maturin experience unfortunately. I can try and have a read/play tomorrow. Maybe we could also ask @bioconda/rust and see if anyone else in that group has Maturin experience, or can point us at a repo that uses it. I think @luizirber might have some experience with Maturin via sourmash??

File: recipes/needletail/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  script:
+    - {{ PYTHON }} -m pip install . --no-deps --no-build-isolation --disable-pip-version-check
+    - cargo-bundle-licenses --format yaml --output THIRDPARTY.yml
+
+requirements:
+  build:
+    - python                              # [build_platform != target_platform]
+    - cross-python_{{ target_platform }}  # [build_platform != target_platform]
+    - crossenv                            # [build_platform != target_platform]
+    - maturin >=1,<2                      # [build_platform != target_platform]
```
Comment (luizirber) [inline recipes/needletail/meta.yaml:26]:
```suggestion
    - maturin >=1,<2                      # [build_platform != target_platform]
    - cffi                                # [build_platform != target_platform]
```

File: recipes/needletail/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    - {{ compiler("c") }}
+    - {{ compiler("rust") }}
+    - {{ stdlib("c") }}
+    - cargo-bundle-licenses
+  host:
+    - python
+    - pip
+    - maturin >=1,<2
```
Comment (luizirber) [inline recipes/needletail/meta.yaml:34]:
```suggestion
    - maturin >=1,<2 
    - cffi
```

**Review by luizirber — CHANGES_REQUESTED** [review]
We moved the sourmash recipe to the v1 format over in https://github.com/conda-forge/sourmash-minimal-feedstock/blob/5f7964f0f0a29b3bf6522a9989cd99c9b4ee3c6b/recipe/recipe.yaml, so some things will be different from here, but I think the only detail missing is `cffi` in the `host` and `build` sections (with comments for cross-compilation)

**Comment by luizirber** [comment]
wait, needletail doesn't use CFFI bindings, it uses pyo3, so maturin is misconfigured...

**Comment by luizirber** [comment]
Added a patch to fix maturin config, might be good to upstream it for next release

**Review by luizirber — APPROVED** [review]

**Comment by mbhall88** [comment]
Amazing @luizirber. Let's see if adding in ARM builds works too 🤞

**Comment by mbhall88** [comment]
@apcamargo I might suggest checking the artifacts to make sure it all works as you expect. When you're happy it does what you expect, feel free to merge.

**Comment by mbhall88** [comment]
@BiocondaBot please fetch artifacts

**Comment by martin-g** [comment]
```
mgrigorov in 🌐 euler-arm-22 in /tmp/needle via 🅒 test123 took 6s 
❯ tar xvf needletail-0.6.1-py310h6ad65a3_0.tar.bz2                                                                                                                                                         (test123) 
info/recipe/test.fa
info/test/test.fa
info/run_exports.json
info/recipe/test.fq
info/test/test.fq
info/hash_input.json
info/recipe/patches/01-maturin-pyo3.patch
info/index.json
info/files
info/licenses/LICENSE
info/recipe/meta.yaml.template
info/paths.json
info/recipe/meta.yaml
info/recipe/run_test.py
info/test/run_test.py
info/recipe/conda_build_config.yaml
info/about.json
info/git
info/licenses/THIRDPARTY.yml
lib/python3.10/site-packages/needletail-0.6.0.dist-info/direct_url.json
lib/python3.10/site-packages/needletail-0.6.0.dist-info/INSTALLER
lib/python3.10/site-packages/needletail-0.6.0.dist-info/WHEEL
lib/python3.10/site-packages/needletail-0.6.0.dist-info/RECORD
lib/python3.10/site-packages/needletail-0.6.0.dist-info/METADATA
lib/python3.10/site-packages/needletail-0.6.0.dist-info/REQUESTED
lib/python3.10/site-packages/needletail/__init__.py
lib/python3.10/site-packages/needletail/__pycache__/__init__.cpython-310.pyc
lib/python3.10/site-packages/needletail/needletail.cpython-310-aarch64-linux-gnu.so
lib/python3.10/site-packages/needletail-0.6.0.dist-info/licenses/LICENSE

mgrigorov in 🌐 euler-arm-22 in /tmp/needle via 🅒 test123 
❯ file lib/**/*.so*                                                                                                                                                                                        (test123) 
lib/python3.10/site-packages/needletail/needletail.cpython-310-aarch64-linux-gnu.so: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, not stripped
```
The Linux-aarch64 build looks good to me!

**Comment by mbhall88** [comment]
I tested out the MAC ARM one and it was good too. Ima merge this. Great work team

**Comment by apcamargo** [comment]
This was quick! Thanks @mbhall88 @luizirber @martin-g!!

**Pull request merged**